### ISSUE TITLE
Using column of joined table as foreign_key:  in has_many: relation

### ISSUE THREAD

**Issue by CroneKorkN** [rails/rails#20597]
```ruby
class AccountTemplate < ActiveRecord::Base
  has_many :accounts
end

class Account < ActiveRecord::Base
  default_scope -> {joins(:account_template)}

  has_many  :closing_here,
             class_name: "Account",
             primary_key: :id,
             foreign_key: "account_templates.close_via_id"
end
```

I want to use a column of the joined table AccountTemplate as foreign key, but ActiveRecord appends the to table name: 
```SQL
WHERE `accounts`.`account_templates.close_via_id` = 2
```
I think AR should generate something like 
```SQL
WHERE `account_templates`.`close_via_id` = 2
```

Please make the foreign_key: param recognize table definitions.

**Comment by simon2k** [comment]
Right, currently it attaches it as a field name. @rafaelfranca should it be handled? I could extend AR to have this ability. :smiley: From the other side, it might have low usage (yagni).

**Comment by CroneKorkN** [comment]
I´m really no rails ninja and i've been struggeling with this problem for so long now. I think many advanced users would appreciate this feature! Using raw SQL is so unelegant.

**Comment by al2o3cr** [comment]
@CroneKorkN - I'm assuming that you have a `belongs_to :account_template` on your `Account` model. In that case, I believe what you want is something like this:
```ruby
class AccountTemplate < ActiveRecord::Base
  has_many :accounts
  belongs_to :close_via, class_name: 'Account'
end

class Account < ActiveRecord::Base
  # not certain you need this, at least for the association
  # default_scope -> {joins(:account_template)}

  belongs_to :account_template
  has_many :closing_here, class_name: 'Account', through: :account_template, source: :close_via
end
```
If this doesn't work as desired, please post the SQL it generated and the SQL you expected.

**Comment by CroneKorkN** [comment]
I am writing a simple accounting app. AccountTemplate defines the abstract relations between the accounts. What closes where and some other attributes.

In the app, you can create independend tasks. Each task has an instance of every Account that is touched by a accounting record. Each Account keeps task-dependent information and belongs_to a AccountTemplate.

```ruby
class AccountTemplate < ActiveRecord::Base
  self.primary_key = :number
  has_many :accounts,
           foreign_key: :number
  belongs_to :close_via,
             class_name:  "AccountTemplate",
             foreign_key: :close_via_number 
  has_many   :closing_here,
             class_name:  "AccountTemplate",
             foreign_key: :close_via_number
end

class Account < ActiveRecord::Base
  belongs_to :task
  belongs_to :account_template,
             foreign_key: :number
  delegate :name,
           :has_initial, 
           :close_via_number,
           to: :account_template
  has_many  :closing_here,
            ?????????????
  belongs_to :close_via,
            ?????????????
end
```
To get the close_via Account of a certain Account, i need to go to its AccountTemplate, then to the AccountTemplates close_via AccountTemplate, from there back to Account-class:

```
Account
-task: 5
-number: 2000
       V
AccountTemplate
-number 2000
-close_via_number 8010
       V
AccountTemplate
-number 8010
-close_via_number nil
       V
Account
-task: id
-number 8010
```

I'm afraid its hard to understand what i want exactly. Never the less i'll keep on playing with "through:", but i'm afraid it won't work, because AccountTemplate doesn't belong_to one Account, but has_many Accounts. Which certain Account is meant, derives from the task_id.

In short: Account belongs_to AccountTemplate. Account belongs_to close_via and has_many closing_here. But the relations between Accounts are defined in the other class, AccountTemplate.
